app/src/main/java/com/github/geohunt/app/data/repository/ClaimRepository.kt

```diff
@@ -0,0 +1,184 @@
+import com.github.geohunt.app.data.network.firebase.models.FirebaseClaim
+import com.github.geohunt.app.model.Challenge
+import com.github.geohunt.app.model.Claim
+import com.github.geohunt.app.model.User
+import com.github.geohunt.app.model.points.GaussianPointCalculator
+import com.github.geohunt.app.model.points.PointCalculator
+import com.github.geohunt.app.utility.DateUtils
+import com.google.firebase.database.FirebaseDatabase
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.awaitAll
+import kotlinx.coroutines.tasks.asDeferred
+import kotlinx.coroutines.tasks.await
+import kotlinx.coroutines.withContext
+
+/**
+ * Contains methods for claim/unclaim a challenge, as well as retrieving the list
+ * of claim for a particular challenge or a particular user
+ */
+class ClaimRepository(
+    private val authRepository: AuthRepository,
+    private val imageRepository: ImageRepository,
+    private val database: FirebaseDatabase = FirebaseDatabase.getInstance(),
+    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO,
+    private val pointCalculatorMap: Map<Challenge.Difficulty, PointCalculator> = mapOf(
+        Challenge.Difficulty.EASY to GaussianPointCalculator(0.20),
+        Challenge.Difficulty.MEDIUM to GaussianPointCalculator(0.15),
+        Challenge.Difficulty.EASY to GaussianPointCalculator(0.10)
+    ).withDefault { GaussianPointCalculator(0.10) },
+) : ClaimRepositoryInterface {
+    private fun getChallengeRefFromId(id: String) {
+        database.getReference("claims/$id")
+    }
+
+    private fun FirebaseClaim.asExternalModel(id: String): Claim = Claim(
+        id = id,
+        parentChallengeId = cid!!,
+        claimerId = uid!!,
+        claimDate = DateUtils.localFromUtcIso8601(time!!),
+        distance = distance,
+        awardedPoints = awardedPoints,
+        photoUrl = photoUrl!!
+    )
+
+    /**
+     * Retrieve a list of all claims id for a specific user, useful when lazy loading
+     */
+    override suspend fun getClaimIdByUser(user: User): List<String> {
+        require(user.id.isNotEmpty())
+
+        return withContext(ioDispatcher) {
+            database.getReference("claimsByUser")
+                .child(user.id)
+                .get()
+                .await()
+                .run {
+                    children.mapNotNull { it.getValue(String::class.java) }
+                }
+        }
+    }
+
+    /**
+     * Check whether the currently logged user claim the given challenges
+     */
+    override suspend fun doesClaims(challenge: Challenge) : Boolean = withContext(ioDispatcher) {
+        authRepository.requireLoggedIn()
+        val currentUser = authRepository.getCurrentUser()
+
+        getClaimsByUser(currentUser)
+            .any { it.parentChallengeId == challenge.id }
+    }
+
+    /**
+     * Retrieve the score for a given user
+     */
+    override suspend fun getScoreFromUser(user: User) : Long {

```

Comment:
I think that `getScore` is clear enough:
```suggestion
    override suspend fun getScore(user: User) : Long {

```